File: github_cpp/16/99.cpp
#include"shellSort.h"

shellSort::shellSort()
{

}
shellSort::shellSort(const shellSort& s)
{

}
shellSort& shellSort::operator=(const shellSort& s)
{

}
shellSort* shellSort::getInstance()
{
	static shellSort instance;
	return &instance;
}

void shellSort::sort(int *p,int sum,int factor)
{
	int d=factor>sum/2?factor:sum/2;
	while(d>0)
	{
		for(int i=d;i<sum;i++)
		{
			int j=i-d;
			while(j>=0&&p[j]>p[j+d])
			{
				int tmp = p[j];
				p[j] = p[j+d];
				p[j+d] = tmp;
				j=j-d;
			}
		}
		
		/*
		for(int i=d;i<sum;i++)
		{
			if(p[i]<p[i-d])
			{
				int j =i-d;
				int x = p[i];
				p[i] = p[i-d];
				while(j>=0&&x<p[j]){
					p[j+d] = p[j];
					j-=d;
				}
				p[j+d] =x;
			}
		}*/

		d= d/2;
	}
}

File: live_test/github_java/sort/7/99.java



public class ShellSort {

    public static void main(String[] args) {
        int[] a = {26,81,12,54,92,76,20,37,19,48};
        ShellSort shellSort = new ShellSort();
        System.out.println("---排序前---");
        shellSort.print(a);
        shellSort.sort(a);
        System.out.println("---排序后---");
        shellSort.print(a);
    }

    private void sort(int[] a) {
        int dk = a.length/2;
        while( dk >= 1  ){
            shellSort(a, dk);
            dk = dk/2;
        }

    }

    private void shellSort(int[] a, int dk) {
        for(int i=dk;i<a.length;i++){
            if(a[i]<a[i-dk]){
                int j;
                int x=a[i];
                a[i]=a[i-dk];
                for(j=i-dk;  j>=0 && x<a[j];j=j-dk){
                    a[j+dk]=a[j];
                }
                a[j+dk]=x;
            }

        }

    }


    private void print(int[] a) {
        for (int i = 0; i < a.length-1; i++) {
            System.out.print(a[i] + " ");
        }
    }
}
